tools/congestion-model/src/strategy/fancy_global_transaction_stop.rs

```diff
@@ -0,0 +1,288 @@
+            return;
+        }
+
+        // We can safely forward the receipt to the receiver.
+        *sent += receipt.attached_gas;
+        ctx.forward_receipt(receipt);
+    }
+
+    // Returns the new transactions gas limit. This is the limit on the chunk
+    // gas beyond which new transactions won't be included. By default it is
+    // GAS_LIMIT but it can be reduced if the outgoing receipt queues are
+    // filling up. Keep in mind that by default incoming receipts are always
+    // included first so if the incoming receipts queue is full then no new
+    // transactions will be included anyway.
+    fn get_new_tx_gas_limit(&self, ctx: &mut ChunkExecutionContext) -> GGas {
+        // Get the max of the outgoing receipts queues across all shards.
+        // This is the same as global transaction stop.
+        let max_gas = self.get_global_max_outgoing_receipts_attached_gas(ctx);
+
+        // How much remaining space is there in the fullest queue.
+        let remaining = self.outgoing_attached_gas_limit.saturating_sub(max_gas);
+
+        // If the fullest queue is less than half full we want to process a full chunk.
+        let half_max = self.outgoing_attached_gas_limit / 2;
+        if remaining > half_max {
+            return GAS_LIMIT;
+        }
+
+        // If the buffer is more than half full scale the GAS_LIMIT.
+        // It's linear with respect to `remaining` with the borders at
+        // * remaining == 0        -> gas limit == 0
+        // * remaining == half_max -> gas_limit == GAS_LIMIT
+        // TODO overflows?
+        GAS_LIMIT * remaining / half_max
+    }
+
+    // Returns the outgoing gas limit. This is how much attached gas we allow
+    // ourselves to send to the other_shard_id. By default it is
+    // max_incoming_attached_gas but it can be reduced if the other shard
+    // incoming receipts queue is filling up.
+    fn get_outgoing_gas_limit(
+        &self,
+        ctx: &mut ChunkExecutionContext,
+        other_shard_id: ShardId,
+    ) -> GGas {
+        let Some(info) = ctx.prev_block_info().get(&other_shard_id) else {
+            return self.outgoing_attached_gas_limit;
+        };
+        let info = info.get::<CongestedShardsInfo>().unwrap();
+
+        let remaining =
+            self.incoming_attached_gas_limit.saturating_sub(info.incoming_receipts_attached_gas);
+
+        // If the queue is less than half full we want to process a full chunk.
+        let half_max = self.incoming_attached_gas_limit / 2;
+        if remaining > half_max {
+            return self.outgoing_attached_gas_limit;
+        }
+
+        // If the buffer is more than half full scale the max_outgoing_attached_gas.
+        // It's linear with respect to `remaining` with the borders at
+        // * remaining == 0        -> gas limit == 0
+        // * remaining == half_max -> gas_limit == max_outgoing_attached_gas
+        // TODO overflows?
+        self.outgoing_attached_gas_limit * remaining / half_max
```

Comment:
I think we are fine regarding overflows as we are using `u64` at Giga granularity.
A PGas is 1_000_000 or roughly 2^20. To run into overflow, we need to square numbers around 2^32 or above, which would be around 4000 PGas. I don't think we plan on such high thresholds, right?
(Or else, you know, just `as u128` it and thank the Rust lang team for giving us an easy solution :wink:  )